Mathlib/Algebra/QuadraticAlgebra/Basic.lean

```diff
@@ -0,0 +1,389 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.Algebra.QuadraticAlgebra.Defs
+import Mathlib.Algebra.Star.Unitary
+import Mathlib.Tactic.FieldSimp
+
+/-! # Quadratic algebras : involution and norm.
+
+Let `R` be a commutative ring. We define:
+
+* `QuadraticAlgebra.star` : the quadratic involution
+
+* `QuadraticAlgebra.norm` : the norm
+
+We prove :
+
+* `QuadraticAlgebra.isUnit_iff_norm_isUnit`:
+  `w : QuadraticAlgebra R a b` is a unit iff `w.norm` is a unit in `R`.
+
+* `QuadraticAlgebra.norm_mem_nonZero_divisors_iff`:
+  `w : QuadraticAlgebra R a b` isn't a zero divisor iff
+  `w.norm` isn't a zero divisor in `R`.
+
+* If `R` is a field, and `∀ r, r ^ 2 ≠ a + b * r`, then `QuadraticAlgebra R a b` is a field.
+`
+-/
+
+namespace QuadraticAlgebra
+
+variable {R : Type*} {a b : R}
+
+section omega
+
+section
+
+variable [Zero R] [One R]
+
+/-- The representative of the root in the quadratic algebra -/
+def ω : QuadraticAlgebra R a b :=
+  ⟨0, 1⟩
+
+@[simp]
+theorem re_omega : (ω : QuadraticAlgebra R a b).re = 0 :=
+  rfl
+
+@[simp]
+theorem im_omega : (ω : QuadraticAlgebra R a b).im = 1 :=
+  rfl
+
+end
+
+variable [CommSemiring R]
+
+theorem omega_mul_omega :
+    (ω : QuadraticAlgebra R a b) * (ω) = ⟨a, b⟩ := by
+  ext <;> simp
+
+theorem omega_mul_omega_eq :
+    (ω : QuadraticAlgebra R a b) * (ω) = a • 1 + b • (ω) := by
+  ext <;> simp
+
+@[simp]
+theorem omega_mul (x y : R) :
+    (ω : QuadraticAlgebra R a b) * ⟨x, y⟩ = ⟨a * y, x + b * y⟩ := by
+  ext <;> simp
+
+@[simp]
+theorem omega_mul_coe_mul (n x y : R) :
+    (ω : QuadraticAlgebra R a b) * (↑n) * ⟨x, y⟩ = ⟨a * n * y, n * x + n * b * y⟩ := by
+  ext <;> simp; ring
+
+theorem omega_prop : (ω) * (ω) =
+    (b : QuadraticAlgebra R a b) * (ω : QuadraticAlgebra R a b) + (a : QuadraticAlgebra R a b) := by
+  ext <;> simp
+
+theorem omega_prop' : (ω : QuadraticAlgebra R a b) * (ω) =
+    b • (ω) + a • 1 := by
+  simp [omega_prop, ← QuadraticAlgebra.coe_mul_eq_smul]
+
+theorem mk_eq_add_smul_omega {x y : R} :
+    (⟨x, y⟩ : QuadraticAlgebra R a b) = x + y • (ω : QuadraticAlgebra R a b) := by
+  ext <;> simp
+
+variable {A : Type*} [Ring A] [Algebra R A]
+
+@[ext]
+theorem hom_ext {f g : QuadraticAlgebra R a b →ₐ[R] A}
+    (h : f (ω) = g (ω)) : f = g := by
+  ext ⟨x, y⟩
+  simp only [mk_eq_add_smul_omega, map_add, map_smul, h, add_left_inj]
+  change f (algebraMap R _ x) = g (algebraMap R _ x)
+  simp only [AlgHom.commutes]
+
+/-- The unique `AlgHom` from `QuadraticAlgebra R a b` to an `R`-algebra `A`,
+constructed by replacing `ω` with the provided root.
+Conversely, this associates to every algebra morphism `QuadraticAlgebra R a b →ₐ[R] A`
+a value of `ω` in `A`. -/
+@[simps]
+def lift : { u : A // u * u = b • u + a • 1 } ≃ (QuadraticAlgebra R a b →ₐ[R] A) where
+  toFun u :=
+    { toFun z := z.re • 1 + z.im • u
```

Comment:
I realize this is more verbose, but I suspect it can simplify some aspects of this PR if you use it in place of `_ • 1` everywhere in the PR.
```suggestion
    { toFun z := algebraMap R (QuadraticAlgebra R a b) z.re + z.im • u
```